### PR TITLE
Add `query` function

### PR DESCRIPTION
**Pull request by philss**
~~This is WIP and may not be merged, but just to leave the idea in the air for now.~~

### CODE REVIEW

**Comment by wojtekmach** [comment]
Looks good to me, by having ReqCH.query(req, …) we can still compose with other pipelines so req in req_ch is justified if that makes sense. :)

**Comment by josevalim** [comment]
If we go this way, we should remove the click house option, right? And how hard would be to adapt req_athena to do the same?

**Comment by philss** [comment]
> If we go this way, we should remove the click house option, right?

I'm not sure. If we decide to keep `attach/2`, then I think we could keep that option. Keeping `attach/2` seems to bring some flexibility, no?

**Comment by josevalim** [comment]
So you are saying that we can skip attach altogether? How would we set and validate the database? Which code would we generate in `kino_db`?

**Comment by wojtekmach** [comment]
One advantage of existing design is we can do this:

```elixir
Req.new() |> ReqCH.attach(clickhouse: sql)
```

that is, build up everything that we need without actually executing it. There is a hidden function (because there are some considerations) called Req.Request.prepare/1 that will run pipeline through all request steps except for adapter, idea was to trivially inspect everything that is _about to_ be sent over the wire.

There is also the following which is also private (because considerations)

    Req.new(plugins: [ReqCH])

(which calls attach/1.)

Not sure it’s super relevant but I thought I’d mention this. Perhaps it’s a case for keeping attach after all.

**Comment by josevalim** [comment]
I doubt the connection will be useful for anything else, given the host and other information will be set for clickhouse.

At the same time, I think we need to have a ReqCH.new or ReqCH.attach anyway, because we need to validate some options. So I think from Req's design point of view, they are pretty equivalent.

The benefit of `ReqCH.query` would be:

* documentation and autocompletion from the user perspective
* less confusing APIs, because when we use `Req.post(clickhouse: ...)`, the `params` option is not used by ReqCH, it attaches its own meaning instead
* `ReqCH.query` can also be used with consistency `Postgrex.query` and friends

Overall, I don't mind going with `ReqCH.new` and `ReqCH.query`. But we would need to do the same changes to `ReqAthena` and release a new version. I think that should be quick though?

**Comment by wojtekmach** [comment]
> `ReqCH.query` can also be used with consistency Postgrex.query and friends

Should it be `ReqCH.query("SELECT number FROM numbers WHERE number > {num:UInt8} LIMIT 3", [num: 5])` then, i.e. separate args like in Postgrex, MyXQL, etc?

That being said the signature would be kinda wild, `ReqCH.query(req \\ Req.new(), sql, params \\ [], opts \\ [])` and not so easy implement with all the defaults. So I think this is fine but I thought I'd bring this up anyway. In hindsight perhaps `Postgrex.query(server, sql | {sql, params}, opts \\ [])` would have been a better API exactly because of defaults but it's obviously moot.

**Comment by josevalim** [comment]
We can easily remove the `req` as an optional argument. I don't see exactly what it is buying us?

**Comment by wojtekmach** [comment]
I suppose the idea was to be consistent with _req_, i.e.: `Req.get(url)`, `Req.get(Req.new(), url: url)`. If we drop it and we have `ReqCH.new`, do we do `ReqCH.query(Req.new(), ...)` or `ReqCH.query(ReqCH.new(), ...)` or both? I'm honestly not sure.

**Comment by josevalim** [comment]
I think we should pick one API to follow. If we try to be consistent with both, we may end-up consistent with none.

**Comment by philss** [comment]
> So you are saying that we can skip attach altogether? How would we set and validate the database? Which code would we generate in kino_db?

@josevalim Sorry for the confusion: I think we can keep `attach/2`, because we can build that `Req.Request` with it, and it makes everything more flexible. For what I understand in the discussion, we could rename `attach/2` and call it `new/2`, right?

> Overall, I don't mind going with ReqCH.new and ReqCH.query. But we would need to do the same changes to ReqAthena and release a new version. I think that should be quick though?

Yes, we can do that there as well. I can change `kino_db` to adhere to the new API.

> Should it be ReqCH.query("SELECT number FROM numbers WHERE number > {num:UInt8} LIMIT 3", [num: 5]) then, i.e. separate args like in Postgrex, MyXQL, etc?

@wojtekmach I would keep as we have today - `{sql, params}`. IMHO it seems easier to understand that a query receives some args, and it's also easier to maintain the API (although not so trivial to decode internally). WDYT?

To summarize, the proposal is the following:

* `new(req, opts \\ []) :: Req.Request.t()` - we could have `new(opts)` too.
* `query(sql_or_query_tuple, opts \\ []) :: Req.Response.t()`.
* `query(req, sql_or_query_tuple, opts \\ [])` :: Req.Response.t() - although we could assume that `opts` here is not necessary if people use `new/2` to build `req`.

WDYT?

**Comment by josevalim** [comment]
No one would ever be able to write this:

```
ReqCH.query("SELECT number FROM numbers WHERE number > {num:UInt8} LIMIT 3", [num: 5])
```

So I don't understand why we are using it as an example. At best you need to pass the base_url, auth, and database, and,  at that point, why not create a Req struct?

**Comment by josevalim** [comment]
Alright, I will make an executive decision for now:

* `ReqCH.new(opts)` - `opts` is the options for `Req.new` + `:database` (which is optional)
* `ReqCH.query(sql, params \\ [], opts \\ [])` (requires the :database option if not given on `new`) - returns `{:ok, req}` or `{:error, ...}`. It returns `{:error, ...}` if not 200
* `ReqCH.query!(sql, params \\ [], opts \\ [])` returns req

There is no need for `attach`. We can always add it later. Let's go with the minimum API for now.

**Comment by wojtekmach** [comment]
I think we also pass :format to new/1 right? And then this should work too:

ReqCH.new(…) |> Req.post!(body: sql)

?

**Comment by josevalim** [comment]
I don’t think we should attach any behavior to Req, besides storing options. We need to decide if people are going to use our API (so we compose) or we inject. I don’t think it makes sense to leave it in a mixed state and support both.

**Comment by josevalim** [comment]
To be clear, we can pass format to new. But it has no use except on ReqCH.query.

**Comment by philss** [comment]
Sorry for the delay. This is ready to another pass :)

File: lib/req_ch.ex [hunk 1]
```diff
@@ -3,103 +3,153 @@ defmodule ReqCH do
...
+  To change that, see the `new/2` docs for details.
   """
 
-  @connection_options [
-    :database
-  ]
-
-  @general_options [
+  @options [
     :clickhouse,
```
Comment (josevalim) [inline lib/req_ch.ex]: I don't think we accept this option anymore?

File: lib/req_ch.ex [hunk 2]
```diff
@@ -3,103 +3,153 @@ defmodule ReqCH do
...
+  """
+  @spec query(sql_query :: binary(), params :: Map.t() | Keyword.t(), opts :: Keyword.t()) ::
+          {:ok, Req.Response.t()} | {:error, binary()}
+  def query(sql_query, params \\ [], opts \\ [])
+
+  def query(sql_query, params, opts)
+      when is_binary(sql_query) and is_query_params(params) and is_list(opts) do
+    opts
+    |> new()
+    |> Req.request(clickhouse: {sql_query, params})
```
Comment (josevalim) [inline lib/req_ch.ex]: Instead of handling it via the clickhouse option, let's just do the query and handle the `:format` directly here. No need to hook into Req's lifecycle. :)

**Review by josevalim — APPROVED** [review]

File: lib/req_ch.ex [hunk 3]
```diff
@@ -3,117 +3,149 @@ defmodule ReqCH do
...
+
+  It's also possible to make a query using `Req.get/2`:
+
+      iex> req = ReqCH.new(database: "system")
+      iex> Req.get!(req, params: [query: "SELECT number FROM numbers LIMIT 3"]).body
+      1\n2\n3\n
+
+  In case the server needs authentication, it's possible to use `Req` options for that.
+
+      iex> req = ReqCH.new(base_url: "https://example.org:8123", auth: {:basic, "user:pass"})
```
Comment (wojtekmach) [inline lib/req_ch.ex]:
```suggestion
      iex> req = ReqCH.new(base_url: "http://example.org:8123", auth: {:basic, "user:pass"})
```

File: lib/req_ch.ex [hunk 4]
```diff
@@ -155,14 +195,11 @@ defmodule ReqCH do
...
   @valid_formats [:tsv, :csv, :json, :explorer]
-  defp normalise_format!(format) do
-    if valid = normalise_format(format) do
-      valid
-    else
-      raise ArgumentError,
-            "the given format #{inspect(format)} is invalid. Expecting one of #{inspect(@valid_formats)} " <>
-              "or one of the valid options described in #{@formats_page}"
-    end
+  defp format_error(format) do
```
Comment (wojtekmach) [inline lib/req_ch.ex]:
```suggestion

  defp format_error(format) do
```

File: lib/req_ch.ex [hunk 5]
```diff
@@ -128,13 +160,21 @@ defmodule ReqCH do
...
-    |> Req.Request.put_private(:clickhouse_format, format)
-    |> Req.Request.put_header("x-clickhouse-format", format_header)
+      request
+      |> Req.Request.put_private(:clickhouse_format, format)
+      |> Req.Request.put_header("x-clickhouse-format", format_header)
+    else
+      Req.Request.halt(
+        request,
+        format_error(format_option)
+      )
```
Comment (wojtekmach) [inline lib/req_ch.ex]: I think this is probably fine but honestly I'd just raise if format is invalid. The way I try to think about it is if user supplies wrong options we want to raise, if the server responds with garbage (e.g. invalid JSON) we gracefully handle that with returning exception struct. (Not that the user can do much with it _anyway_ but people wanted that so they got it.)

File: lib/req_ch.ex [hunk 6]
```diff
@@ -3,117 +3,149 @@ defmodule ReqCH do
...
       #Explorer.DataFrame<
         Polars[3 x 1]
         number u64 [0, 1, 2]
       >
 
   Using parameters is also possible:
 
-      iex> req = Req.new() |> ReqCH.attach(format: :explorer, database: "system")
-      iex> Req.get!(req, clickhouse: {"SELECT number FROM numbers WHERE number > {num:UInt8} LIMIT 3", [num: 5]}).body
+      iex> opts = [format: :explorer, database: "system"]
```
Comment (wojtekmach) [inline lib/req_ch.ex:103]: I think it'd be nice to show passing "regular" Req options that are important for CH like :base_url and :basic so it's clear that both ReqCH.new+Req.post and ReqCH.query are equivalent.

**Review by wojtekmach — APPROVED** [review]
Just some minor things left.

File: lib/req_ch.ex (hunk 6 above)
Comment (philss) [reply lib/req_ch.ex:103]: I added one more example showing that. WDYT?

File: lib/req_ch.ex (hunk 5 above)
Comment (philss) [reply lib/req_ch.ex]: Yeah, I was unsure about this, so I changed to raise instead. Thanks!

File: lib/req_ch.ex (hunk 6 above)
Comment (wojtekmach) [reply lib/req_ch.ex:103]: Yup sounds good!
